File: Doxygen2HugoConverter/Doxygen2HugoConverter/Generator/ClassGenerator.fs
module ClassGenerator

open System.IO
open System.Text

let processFields (context: GeneratorCommon.Context) (classDef: Defs.ClassDef) (dest: StringBuilder) =
    classDef.Fields |> Seq.iter (fun fieldDef -> fieldDef |> FieldGenerator.generate context)
    match classDef.MemberRefs |> FieldGenerator.getFieldEntries context with
    | [] -> ()
    |fieldEntries ->
        GeneratorCommon.generateHeader "Fields" 2 |> dest.Append |> ignore
        dest |> GeneratorCommon.generateTableHeader ["Field"; "Description"]
        fieldEntries |> Seq.iter (fun entry -> sprintf $"| {entry.Title} | {entry.BriefDescription} |" |> dest.AppendLine |> ignore)

let processMethods (context: GeneratorCommon.Context) (classDef: Defs.ClassDef) (dest: StringBuilder) =
    classDef.DirectMethods |> Seq.iter (fun group -> group |> MethodGenerator.generateForMethodGroup context)
    match classDef.MemberRefs |> MethodGenerator.getMethodEntries context with
    | [] -> ()
    | methodEntries ->
        GeneratorCommon.generateHeader "Methods" 2 |> dest.Append |> ignore
        dest |> GeneratorCommon.generateTableHeader ["Method"; "Description"]
        methodEntries |> Seq.iter (fun entry -> sprintf $"| {entry.Title} | {entry.BriefDescription} |" |> dest.AppendLine |> ignore)

let processTypedefs (context: GeneratorCommon.Context) (classDef: Defs.ClassDef) (dest: StringBuilder) =
    classDef.Typedefs |> Seq.iter (fun field -> field |> TypedefGenerator.generate context |> ignore)
    match classDef.Typedefs |> TypedefGenerator.createTypedefEntries context with
    | [] -> ()
    | typedefEntries ->
        GeneratorCommon.generateHeader "Typedefs" 2 |> dest.Append |> ignore
        dest |> GeneratorCommon.generateTableHeader ["Typedef"; "Description"]
        typedefEntries |> Seq.iter (fun entry -> sprintf $"| {entry.Title} | {entry.BriefDescription} |" |> dest.AppendLine |> ignore)

let generateKind (classDef: Defs.ClassDef) =
    match classDef.Kind with
    | Defs.ClassKind.Class -> "class"
    | Defs.ClassKind.Interface -> "interface"

let generateBaseClass (baseClassDef: Defs.BaseClassDef) =
    let result = new StringBuilder()
    baseClassDef.Access |> result.Append |> ignore
    " "|> result.Append |> ignore
    if baseClassDef.Virtual then
        "virtual "|> result.Append |> ignore
    baseClassDef.QualifiedName.Replace("< ", "<").Replace(" >", ">") |> result.Append |> ignore
    result.ToString()

let generateDefinition (classDef: Defs.ClassDef) (dest: StringBuilder) =
    "```cpp" |> dest.AppendLine |> ignore
    if classDef.TemplateParameters.IsEmpty |> not then
        "template<" |> dest.Append |> ignore
        classDef.TemplateParameters |> String.concat "," |> dest.Append |> ignore
        "> " |> dest.Append |> ignore
    "class " |> dest.Append |> ignore
    classDef.Name |> dest.Append |> ignore
    if classDef.BaseClasses.IsEmpty |> not then
        " : " |> dest.Append |> ignore
        classDef.BaseClasses |> Seq.map generateBaseClass |> String.concat ", " |> dest.Append |> ignore
    dest.AppendLine() |> ignore
    "```" |> dest.AppendLine |> ignore
    dest.AppendLine() |> ignore

let generate (context: GeneratorCommon.Context) (classDef: Defs.ClassDef) =
    let folderName = classDef.Name |> Utils.createSimpleFolderName
    let classDirectory = Path.Combine(context.Directory, folderName)
    classDirectory |> Directory.CreateDirectory |> ignore
    let classUrl = [|folderName|] |> Array.append context.Url
    let currentContext = {context with Directory = classDirectory; Url = classUrl; Weight = 1}
    let urlGenerator = GeneratorUrl.generateRelativeUrlForEntity currentContext
    let briefDescriptionGenerator = MarkupGenerator.GenerateBriefDescription urlGenerator
    let builder = new StringBuilder()
    let descriptionForTitle = classDef.BriefDescription |> MarkupGenerator.GenerateBriefDescriptionForTitle
    builder |> GeneratorCommon.generateDefPageHeader classDef.Name descriptionForTitle classUrl context.Weight
    context.Weight <- context.Weight + GeneratorCommon.WeightDelta
    GeneratorCommon.generateHeader (sprintf $"{classDef.Name} {classDef |> generateKind}") 2 |> builder.Append |> ignore
    let briefDescription = classDef.BriefDescription |> briefDescriptionGenerator
    builder.AppendLine() |> ignore
    briefDescription |> builder.AppendLine |> ignore
    builder.AppendLine() |> ignore
    builder |> generateDefinition classDef
    builder |> GeneratorCommon.generateTemplateParameters briefDescriptionGenerator classDef.DetailedDescription.TemplateParameters
    builder |> processMethods currentContext classDef
    builder |> processFields currentContext classDef
    builder |> processTypedefs currentContext classDef
    let detailedDescription = classDef.DetailedDescription.Description |> MarkupGenerator.GenerateDetailedDescription urlGenerator
    detailedDescription |> builder.Append |> ignore
    File.AppendAllText(Path.Combine(classDirectory, Common.MarkdownFilename), builder.ToString())

let createEntry (briefDescriptionGenerator: Markup.SimpleMarkup -> string) (classDef: Defs.ClassDef) =
    let folderName = classDef.Name |> Utils.createSimpleFolderName
    let briefDescription = classDef.BriefDescription |> briefDescriptionGenerator
    {GeneratorCommon.GenerateEntry.Title = folderName |> GeneratorCommon.generateChildUrl |> GeneratorCommon.generateLink classDef.Name;
     GeneratorCommon.GenerateEntry.BriefDescription = briefDescription}

let createClassEntries (context: GeneratorCommon.Context) (classDefs: Defs.ClassDef list) =
    let urlGenerator = GeneratorUrl.generateRelativeUrlForEntity context
    let briefDescriptionGenerator = MarkupGenerator.GenerateBriefDescription urlGenerator
    classDefs |> Seq.map (fun classDef -> classDef |> createEntry briefDescriptionGenerator) |> Seq.toList

File: Doxygen2HugoConverter/Doxygen2HugoConverter/Generator/ClassGenerator.cs
using System.Text;
using Doxygen2HugoConverter.Entities;
using Doxygen2HugoConverter.Lookup;

namespace Doxygen2HugoConverter.Generator
{
    internal static class ClassGenerator
    {
        public static void GenerateForClass(this EntityDef.ClassEntity entity, GenerateState state, LookupFrame currentFrame)
        {
            String folderName = NameUtils.CreateSimpleFolderName(entity.Name);
            String classDirectory = Path.Combine(state.Directory, folderName);
            Directory.CreateDirectory(classDirectory);
            IList<String> classUrl = state.Url.Append(folderName).ToList();
            GenerateState currentState = new GenerateState(classDirectory, classUrl, state.ConvertData);
            String? CreateUrl(String entityId) => UrlGenerator.CreateRelativeUrlForEntity(entityId, currentState);
            StringBuilder builder = new StringBuilder();
            String classKind = entity.GetClassKind();
            String defaultTitleDescription = entity.CreateDefaultHeaderDescription(state.ConvertData);
            String descriptionForTitle = entity.BriefDescription.CreateBriefDescriptionForTitle($"{entity.FullName} {classKind}", defaultTitleDescription);
            GeneratorUtils.GenerateDefPageHeader($"{entity.FullName} {classKind}", entity.Name, descriptionForTitle, classUrl, currentFrame.CurrentWeight, state.ConvertData, builder);
            GeneratorUtils.GenerateHeader($"{entity.Name} {classKind}", 2, builder);
            String briefDescription = entity.BriefDescription.CreateSimpleMarkup(CreateUrl, currentState.ConvertData.Logger);
            builder.AppendLine();
            builder.AppendLine(briefDescription);
            builder.AppendLine();
            entity.GenerateClassDefinition(builder);
            entity.DetailedDescription.TemplateParameters.GenerateTemplateParameters(CreateUrl, builder, currentState.ConvertData.Logger);
            String[] entities = entity.DirectMethods
                .Select(group => group.Name)
                .Union(entity.Fields.Select(field => field.Name))
                .Union(entity.Typedefs.Select(typedef => typedef.Name))
                .ToArray();
            currentFrame.FillKnownChildren(entities);
            entity.ProcessMethods(currentState, currentFrame, builder);
            entity.ProcessFields(currentState, currentFrame, builder);
            entity.ProcessTypedefs(currentState, currentFrame, builder);
            entity.DetailedDescription.Description.GenerateDetailedDescription(CreateUrl, builder, currentState.ConvertData.Logger);
            entity.GenerateSeeAlso(currentState, builder);
            File.AppendAllText(Path.Combine(classDirectory, Common.MarkdownFilename), builder.ToString());
        }

        public static IList<GenerateEntry> CreateClassEntries(this IList<EntityDef.ClassEntity> entities, GenerateState state)
        {
            String? CreateUrl(String entityId) => UrlGenerator.CreateRelativeUrlForEntity(entityId, state);
            GenerateEntry? CreateEntry(EntityDef.ClassEntity entity)
            {
                String folderName = NameUtils.CreateSimpleFolderName(entity.Name);
                String title = GeneratorUtils.CreateLink(entity.Name, UrlGenerator.CreateChildUrl(folderName));
                String briefDescription = entity.BriefDescription.CreateSimpleMarkup(CreateUrl, state.ConvertData.Logger);
                return new GenerateEntry(title, briefDescription);
            }
            return entities.Choose(CreateEntry).ToList();
        }

        private static String CreateBaseClass(this BaseClassEntity entity)
        {
            StringBuilder result = new StringBuilder();
            result.Append($"{entity.Access} ");
            if (entity.Virtual)
                result.Append("virtual ");
            result.Append(entity.QualifiedName.Replace("< ", "<").Replace(" >", ">"));
            return result.ToString();
        }

        private static void GenerateBaseClasses(this IList<BaseClassEntity> entities, Int32 indentationValue, StringBuilder dest)
        {
            dest.AppendLine($"{entities[0].CreateBaseClass()}{(entities.Count > 1 ? "," : "")}");
            String indentation = new String(' ', indentationValue);
            Int32 maxIndex = entities.Count - 1;
            for (Int32 index = 1; index < entities.Count; ++index)
                dest.AppendLine($"{indentation}{entities[index].CreateBaseClass()}{(index < maxIndex ? "," : "")}");
        }

        private static void GenerateClassDefinition(this EntityDef.ClassEntity entity, StringBuilder dest)
        {
            dest.AppendLine("```cpp");
            Int32 startSize = dest.Length;
            if (!entity.TemplateParameters.IsEmpty())
                dest.Append($"template<{String.Join(',', entity.TemplateParameters)}>");
            dest.Append($"class {entity.Name}");
            if (entity.BaseClasses.IsEmpty())
                dest.AppendLine();
            else
            {
                dest.Append(" : ");
                Int32 indentationValue = dest.Length - startSize;
                entity.BaseClasses.GenerateBaseClasses(indentationValue, dest);
            }
            dest.AppendLine("```");
            dest.AppendLine();
        }

        private static void ProcessMethods(this EntityDef.ClassEntity entity, GenerateState state, LookupFrame currentFrame, StringBuilder dest)
        {
            entity.DirectMethods.GenerateForChildren(currentFrame, (group, childFrame) => group.GenerateForMethodGroup(state, childFrame), group => group.Name);
            IList<GenerateEntry> entries = entity.MemberRefs.GetMethodEntries(state);
            if (entries.IsEmpty())
                return;
            GeneratorUtils.GenerateHeader("Methods", 2, dest);
            GeneratorUtils.GenerateTableHeader(new[]{"Method", "Description"}, dest);
            entries.Iterate(entry => { dest.AppendLine($"| {entry.Title} | {entry.BriefDescription} |"); });
        }

        private static void ProcessFields(this EntityDef.ClassEntity entity, GenerateState state, LookupFrame currentFrame, StringBuilder dest)
        {
            entity.Fields.GenerateForChildren(currentFrame, (fieldEntity, childFrame) => fieldEntity.GenerateForField(state, childFrame));
            IList<GenerateEntry> entries = entity.MemberRefs.GetFieldEntries(state);
            if (entries.IsEmpty())
                return;
            GeneratorUtils.GenerateHeader("Fields", 2, dest);
            GeneratorUtils.GenerateTableHeader(new[] {"Field", "Description"}, dest);
            entries.Iterate(entry => { dest.AppendLine($"| {entry.Title} | {entry.BriefDescription} |"); });
        }

        private static void ProcessTypedefs(this EntityDef.ClassEntity entity, GenerateState state, LookupFrame currentFrame, StringBuilder dest)
        {
            entity.Typedefs.GenerateForChildren(currentFrame, (typedefEntity, childFrame) => typedefEntity.GenerateForTypedef(state, childFrame));
            IList<GenerateEntry> entries = entity.Typedefs.GetTypedefEntries(state);
            if (entries.IsEmpty())
                return;
            GeneratorUtils.GenerateHeader("Typedefs", 2, dest);
            GeneratorUtils.GenerateTableHeader(new[] {"Typedef", "Description"}, dest);
            entries.Iterate(entry => { dest.AppendLine($"| {entry.Title} | {entry.BriefDescription} |"); });
        }

        private static void GenerateSeeAlso(this BaseClassEntity baseClassEntity, GenerateState state, StringBuilder dest)
        {
            String? CreateUrl(String entityId) => UrlGenerator.CreateRelativeUrlForEntity(entityId, state);
            String refId = baseClassEntity.RefId!;
            switch (state.ConvertData.EntityRepo.ContainsKey(refId))
            {
                case false:
                    state.ConvertData.Logger.LogWarning($"Can't resolve ref with id = \"{refId}\"");
                    break;
                case true:
                    String url = CreateUrl(refId)!;
                    state.ConvertData.Logger.LogInfo($"Resolved ref for \"{baseClassEntity.QualifiedName}\": \"{url}\"");
                    state.ConvertData.EntityRepo[refId].GenerateSeeAlsoEntry(url, dest);
                    break;
            }
        }

        private static void GenerateSeeAlso(this EntityDef.ClassEntity entity, GenerateState state, StringBuilder dest)
        {
            GeneratorUtils.GenerateHeader("See Also", 2, dest);
            entity.BaseClasses
                  .Where(baseClassEntity => !String.IsNullOrEmpty(baseClassEntity.RefId))
                  .DistinctBy(baseClassEntity => baseClassEntity.RefId)
                  .Iterate(baseClassEntity => baseClassEntity.GenerateSeeAlso(state, dest));
            entity.GenerateSeeAlsoCommonPart(state.ConvertData.EntityRepo, state.ConvertData, dest);
        }

        private static String GetClassKind(this EntityDef.ClassEntity entity)
        {
            return entity.Kind switch
            {
                ClassKind.Class => "class",
                ClassKind.Interface => "interface",
                _ => throw new ArgumentOutOfRangeException(nameof(entity.Kind), entity.Kind, null)
            };
        }
    }
}
